Store.js

```diff
@@ -140,14 +151,130 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+					this.observeCallbackItems = this._observeCallbackItems.bind(this);
+					for (var i = 0; i < this.store.length; i++) {
+						// affect the callback to the observe function if the item is observable
+						this.itemHandles[i] = Observable.observe(this.store[i], this.observeCallbackItems);
+					}
+					collection = processQueryResult.call(this, this.store.filter(this._isQueried, this));
+					// affect the callback to the observe function if the array is observable
+					this.storeHandle = ObservableArray.observe(this.store, this.observeCallbackArray);
 				}
+				return this.processCollection(collection);
 			} else {
 				this.initItems([]);
 			}
 		},
 
 		/**
+		 * Synchronously deliver change records to all listeners registered via `observe()`.
+		 */
+		deliver: dcl.superCall(function (sup) {
+			return function () {
+				sup.call();
+				if (this.storeHandle) {
+					this.storeHandle.deliver();
+				}
+				if (this.itemHandles) {
+					this.observeCallbackItems && Observable.deliverChangeRecords(this.observeCallbackItems);
+				}
+			};
+		}),
+
+		/**
+		 * Discard change records for all listeners registered via `observe()`.
+		 */
+		discardChanges: dcl.superCall(function (sup) {
+			return function () {
+				sup.call();
+				if (this.storeHandle && this.itemHandles) {
+					this._beingDiscarded = true;
+					this.storeHandle.deliver();
+					this.observeCallbackItems && Observable.deliverChangeRecords(this.observeCallbackItems);
+					this._beingDiscarded = false;
+					return this.store;
+				}
+			};
+		}),
+
+		/**
+		 * Called to verify if the item respect the query conditions
+		 * @param item
+		 * @private
+		 */
+		_isQueried: function (item) {
+			for (var prop in this.query) {
+				if (item[prop] !== this.query[prop]) {
+					return false;
+				}
+			}
+			return true;
+		},
+
+		/**
+		 * Called when a modification is done on the array.
+		 * @param changeRecords - send by the Observe function
+		 */
+		_observeCallbackArray: function (changeRecords) {
+			if (!this._beingDiscarded) {
+				for (var i = 0; i < changeRecords.length; i++) {
+					if (changeRecords[i].type === "splice") {
+						var j;
+						for (j = 0; j < changeRecords[i].removed.length; j++) {
+							this.itemHandles[changeRecords[i].index].remove();
+							var evtRemoved = {previousIndex: changeRecords[i].index};
+							this._itemRemoved(evtRemoved);
+						}
+						for (j = 0; j < changeRecords[i].addedCount; j++) {
+							var evtAdded = {
+								index: changeRecords[i].index + j,
+								target: this.store[changeRecords[i].index + j]
+							};
+							if (this.renderItems !== null && this.renderItems !== undefined) {
+								evtAdded.index = evtAdded.index <= this.renderItems.length ?
+									evtAdded.index : this.renderItems.length;
+							}
+							// affect the callback to the observe function if the item is observable
+							this.itemHandles[changeRecords[i].index + j] =
```

Comment:
Would you want to splice in the new handle instead of (possibly) overwriting the handle of given index?
Also please make sure unit test covers handle cleanup stuffs, in case it hasn’t already.